clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/gitRepo/GitRepoArtifactCredentials.java

```diff
@@ -0,0 +1,150 @@
+
+  public GitRepoArtifactCredentials(GitRepoArtifactAccount account) {
+    this.name = account.getName();
+    this.username = account.getUsername();
+    this.password = account.getPassword();
+  }
+
+  @Override
+  public InputStream download(Artifact artifact) throws IOException {
+    String repoReference = artifact.getReference();
+    Path stagingPath =
+        Paths.get(System.getProperty("java.io.tmpdir"), UUID.randomUUID().toString());
+
+    try {
+      log.info("Cloning git/repo {} into {}", repoReference, stagingPath.toString());
+      Git localRepository = clone(artifact, stagingPath);
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      log.info("Creating archive for git/repo {}", repoReference);
+      archiveToOutputStream(artifact, localRepository, outputStream);
+      return new ByteArrayInputStream(outputStream.toByteArray());
+    } catch (Exception e) {
+      throw new IOException("Failed to clone git/repo " + repoReference, e);
+    } finally {
+      log.info("Cleaning up git/repo {}", repoReference);
+      try {
+        FileUtils.deleteDirectory(stagingPath.toFile());
+      } catch (IOException e) {
+        log.info("Failed to cleanup git/repo for {}", repoReference);
+        throw new RuntimeException(
+            "Failed to clean up git/repo staging directory for repo" + repoReference, e);
+      }
+    }
+  }
+
+  private Git clone(Artifact artifact, Path stagingPath) throws Exception {
+    String version = artifact.getVersion() != "" ? artifact.getVersion() : "master";
+    String subPath = artifactSubPath(artifact);
+    // TODO(ethanfrogers): add support for clone history depth once jgit supports it
+
+    Git localRepository =
+        addAuthentication(Git.cloneRepository())
+            .setURI(artifact.getReference())
+            .setDirectory(stagingPath.toFile())
+            .setNoCheckout(true)
+            .call();
+
+    CheckoutCommand checkoutCommand =
+        localRepository.checkout().setName(version).setStartPoint("origin/" + version);
+
+    if (!StringUtils.isEmpty(subPath)) {
+      checkoutCommand = checkoutCommand.addPath(subPath);
+    }
+
+    checkoutCommand.call();
+    return localRepository;
+  }
+
+  private void archiveToOutputStream(Artifact artifact, Git repository, OutputStream outputStream)
+      throws Exception {
+    String version = artifact.getVersion() != "" ? artifact.getVersion() : "master";
+    String subPath = artifactSubPath(artifact);
+
+    try {
+      ArchiveFormats.registerAll();
+      ArchiveCommand archiveCommand =
+          repository
+              .archive()
+              .setTree(repository.getRepository().resolve("origin/" + version))
+              .setFormat("tgz")
+              .setOutputStream(outputStream);
+
+      if (!StringUtils.isEmpty(subPath)) {
+        archiveCommand.setPaths(subPath);
+      }
+
+      archiveCommand.call();
+    } finally {
+      ArchiveFormats.unregisterAll();
```

Comment:
I think calling `registerAll()` followed by `unregisterAll()` each time we archive a repo is going to introduce threading bugs, as this is updating static variables that are used by all threads.  I suspect we really just want to call `registerAll()` (or even a more specific register of `tgz`) just once on startup.

Reply:
Ah, glad you called that out. I've removed the calls to `register/unregister` and added a call to register the single `tgz` format when the credentials are constructed. It will get re-registered for every account but that seems like a fine tradeoff to me. The alternative is that the `JGit` implementation leaks into other areas of this. I tried to register only if unregistered but there isn't an API for determining what types are already registered.

Reply:
It looks like the `registerFormat` does reference counting, so maybe we don't want to create one on every account...perhaps this would be better in a static initializer?  Though honestly I think it's probably fine where it is too.

The javadoc suggests calling `registerFormat` then `unregisterFormat` on each use, but it definitely seems that it's not threadsafe so I think what you have here or a static initializer is definitely better.